client/src/legacy/GridField.js

```diff
@@ -180,6 +194,36 @@ $.entwine('ss', function($) {
       if (hasLegacyFilterHeader) {
         this.find('.sortable-header th:last').html(content);
       }
+    },
+
+    keepStateInHistory: function() {
+      const newURLScheme = $(this).find('.gridfield-actionmenu__container').data('schema');
+      if (newURLScheme && newURLScheme.length > 0) {
+        newURLScheme.filter( e => {
+          if (e.type === 'link') {
+            const reqString = this.buildURLString(e.url)
+            const searchParam =  reqString ? '?' + reqString.join('&') : '';
+            const historyState = $.extend({}, {path: window.location.pathname + searchParam}, this.getState());
+            history.replaceState(historyState, '', window.location.pathname + searchParam);
+          }
+        })
+      }
+    },
+
+    buildURLString: function(url) {
+      const link = [window.location.origin, url].join('/');
+      const params = [window.location.search, (new URL(link)).searchParams.toString()].join('&').substring(1);
+      const newrequest = [];
+      const reqString = [];
+      params.split('&').forEach(param => {
+        const newVal = param.split('=');
+        newrequest[newVal[0]] = newVal[1] ? newVal[1] : '';
+      });
+      Object.keys(newrequest).forEach(param => {
+        reqString.push([param, newrequest[param]].join('='));
+      });
```

Comment:
Can you please explain what this is doing? We should already have a valid query string before you do any of this.

Reply:
This just clean up request string from identical keys and leaves only new one. Otherwise, we just concatenate new state for some key to existing GET request that already has this key. 
Like this : ?**test**=3&**gridState-SilverStripe-FrameworkTest-Model-Company-0**=%7B"GridFieldSortableHeader"%3A%7B"SortColumn"%3A"Name"%7D%7D&**gridState-SilverStripe-FrameworkTest-Model-Company-0**=%7B"GridFieldSortableHeader"%3A%7B"SortColumn"%3A"Name"%2C"SortDirection"%3A"desc"%7D%7D

Reply:
Probably just add a comment above this that explains you're removing duplicate entries. It's a bit hard to understand that at a glance.

Reply:
DONE